### ISSUE TITLE
Add ability to dynamically limit graph depth in `Formatter`

### ISSUE THREAD

**Issue by shatl** [fluentassertions/fluentassertions#1274]
Add ability to limit graph depth in `Formatter`

Currently it is hardcoded to *5*,  which makes error messages for assertions on collections containing complex object graphs unreadable. 

```csharp
 private static string FormatChild(string path, object childValue, bool useLineBreaks, ObjectGraph graph) {
    /// ...
                else if (graph.Depth > 5)
                {
                    return "{Maximum recursion depth was reached…}";
                }
}
```

Would be nice to have an option to pass depth to collection assertion e.g. 
```csharp
result.Should().Contain(p => p.Foo == "Bar", maxDepth: 1);
```

### Current behavior

For list of 2 complex objects, I'm getting message `Expected result { ...` containing almost 200 lines of text

### Versions

* FluentAssertions: 5.10.2 
* .NET Core 2.2.

**Comment by dennisdoomen** [comment]
> ```cs
> result.Should().Contain(p => p.Foo == "Bar", maxDepth: 1);
> ```

I understand your problem, but this would pollute the entire set of APIs. Maybe a global setting would work better.

**Comment by jnyrup** [comment]
How about adding a `(Concurrent?)Dictionary<Type, int>` to `Formatter` to allow specifying the `maxDepth` per type?

```c#
private static Dictionary<Type, int> MaxDepths { get; }

public static int MaxDepth { get; set; } = 5;

public static SetMaxDepth<T>(int maxDepth) =>
    MaxDepths[typeof(T)] = maxDepth;

private int GetMaxDepth(object obj) => 
    MaxDepths.TryGetValue(obj.GetType(), out int maxDepth)
    ? maxDepth
    : MaxDepth;

...
else if (graph.Depth > GetMaxDepth(childValue))
```

**Comment by jnyrup** [comment]
@shatl If you always want to limit the formatting of your complex type, you can also add a [custom formatter](https://fluentassertions.com/extensibility/#rendering-objects-with-beauty) that only contains the relevant properties and add it using `Formatter.AddFormatter`.

**Comment by shatl** [comment]
@dennisdoomen  Global setting should work, assuming we are not expecting our tests to always fail :) 

The question is how and where to set it for projects with big number of unit-tests.

**Comment by jnyrup** [comment]
@shatl Where to set it depends on which unit test framework you're using.
E.g. MSTest has [`[AssemblyInitialize]`](https://docs.microsoft.com/en-us/previous-versions/visualstudio/visual-studio-2008/ms245278(v%3Dvs.90)#examples).

In any case I *wouldn't* put directly in a test method as it alters global state.

**Comment by pgolinski** [comment]
> @shatl If you always want to limit the formatting of your complex type, you can also add a [custom formatter](https://fluentassertions.com/extensibility/#rendering-objects-with-beauty) that only contains the relevant properties and add it using `Formatter.AddFormatter`.

Adding value formatter just to limit graph depth is not a solution. It's a workaround for me.

**Comment by jnyrup** [comment]
@pgolinski It is indeed a workaround.
I mentioned it because sometimes a good workaround will do until a solution might be available.

Would my sketch above with `MaxDepth` and `SetMaxDepth` be a useful solution to your case?

**Comment by pgolinski** [comment]
Yes. It would be easy to configure for simple cases. 

I must confess I used your workaround to cut off ILog from log4net ;). Thank you.

```        
        private class LogFormatter : IValueFormatter
        {
            public bool CanHandle(object value) => value is ILog;

            public string Format(object value, FormattingContext context, FormatChild formatChild)
                => value.ToString();
        }
```

**Comment by dennisdoomen** [comment]
This is resolved in #1469.
